### ISSUE TITLE
Too many open files with large number of small tasks

### ISSUE THREAD

**Issue by azjps** [dask/distributed#1858]
Running on `python==2.7.6, dask==0.17.1, distributed==1.21.4, tornado==4.5.3`.

```python
import time
from distributed import as_completed
from distributed import Client, LocalCluster

def local_func(x):
  time.sleep(0.2)
  return x

if __name__ == '__main__':
  cluster = LocalCluster(n_workers=40)
  client = Client(cluster)

  futures = []
  for i in range(4000):
    futures.append(client.submit(local_func, i)) 

  futures2 = []
  for future, result in as_completed(futures, with_results=True):
    futures2.append(client.submit(local_func, result))

  client.gather(futures)
  # print([fut.result() for fut in futures])
```

After executing this script and watching `lsof -p`, the open file descriptor count quickly jumps to 4000+ (and in my case quickly hits `ulimit -n`). The maximum file descriptor count appears to be somewhat non-deterministic and appears to depends on the various hard-coded values above (shorter sleep, more workers, and more tasks all seem to help make this condition happen). Either way, I wouldn't expect this code to hit any file descriptor limit, since the number of workers is small.

The following exceptions get logged repeatedly:
```bash
tornado.application - ERROR - Exception in callback <bound method SystemMonitor.update of <SystemMonitor: cpu: 120 memory: 319 MB fds: 4037>>
Traceback (most recent call last):
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/tornado/ioloop.py", line 1029, in _run
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/distributed/system_monitor.py", line 49, in update
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/__init__.py", line 1060, in cpu_percent
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/_pslinux.py", line 1365, in wrapper
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/_pslinux.py", line 1521, in cpu_times
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/_common.py", line 336, in wrapper
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/_pslinux.py", line 1404, in _parse_stat_file
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/psutil/_pslinux.py", line 190, in open_binary
IOError: [Errno 24] Too many open files: '/proc/1988210/stat'
tornado.application - ERROR - Exception in callback (<socket._socketobject object at 0x4638600>, <function null_wrapper at 0x463a7d0>)
Traceback (most recent call last):
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/tornado/ioloop.py", line 891, in start
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/tornado/stack_context.py", line 277, in null_wrapper
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/site-packages/tornado/netutil.py", line 264, in accept_handler
  File "/apps/stonefs1/stonepy/venv/20180222.el6-stonelib/lib/python2.7/socket.py", line 202, in accept
error: [Errno 24] Too many open files
```

(Note: I came across this while using `dask_drmaa` and not `LocalCluster`. It is not typical for me to submit lots of very short [e.g. 100ms-1s range] jobs -- I came across this accidentally -- but I also wouldn't expect shorter tasks to cause a crash.)

**Comment by mrocklin** [comment]
It looks like each nanny/worker takes around 8 or 9 file descriptors to start up.  There will then be more between them if there is inter-worker communication.  I would not expect any difference between short and long-running tasks.  You say that you experience different behavior if you change the sleep time above?

**Comment by azjps** [comment]
Yes, the above script finishes fine for me if say I change the sleep time to 1 second instead of 0.2 seconds.

To clarify, I don't think its related to the number of file descriptors used by each nanny/worker, but rather due to calls to `psutil.Process.cpuinfo()/memory_info()` in periodic callbacks, namely these lines:

```python
  File "distributed/system_monitor.py", line 49, in update
  File "distributed/nanny.py", line 254, in memory_monitor
```

Commenting these periodic callbacks locally does seem to avoid this issue, although I'm afraid I don't have much insight as to what's going on. Could these periodic callbacks be getting queued and then run simultaneously?

**Comment by mrocklin** [comment]
Hrm, the fact that commenting those out solves the problem is quite interesting.  Those functions do open a system file through the `psutil` calls, though only very briefly and they close it directly afterwards.

I would expect Tornado's PeriodicCallbacks to handle avoiding the queuing problem, but it would not be hard to add a small bit of code in there to check about the last time called to verify this if you're interested.

**Comment by carlfischerjba** [comment]
Has anyone come up with a workaround for this issue? I'm encountering the same error when I use `map_partitions()` on a `dask.dataframe` with 1M rows. Using a 100k row dataframe runs without errors. The snippet above causes the same error as reported.

I'm still working to identify exactly which piece of my code triggers this behaviour but I expect it will be something like this:
```python
cluster = LocalCluster()
client = Client(cluster)

data2_list = []
for group, dataset_for_group in dataset.groupby('group'):
    dataset_for_group_chunked = dd.from_pandas(dataset_for_group, chunksize=chunksize)
    data2_list.append(dataset_for_group_chunked.map_partitions(func2, group, param1, param2))

data2 = dd.concat(data2_list, interleave_partitions=True).compute()
```

**Comment by carlfischerjba** [comment]
After commenting out other parts of code intertwined with the snippet above, I can confirm that's pretty much the code that triggers the error, except that it's actually running `compute(get=get_sync)`.

My `func2()` opens and reads values from a few rasters (one at a time, and closes them afterwards) and having multiple processes doing this in parallel slows it down a lot. I'm not actually sure that `get_sync` achieves single process execution because I still see multiple processes. I'll try `compute(num_workers=1)` instead and see how this affects performance and the "too many open files" error.

**Comment by bachsh** [comment]
I'm still getting this when running many short tasks (many=100). Moreover, after the job fails, I cannot submit any more tasks to the scheduler, getting the same error right away. It appears that file descriptors stay open for some reason. I'm using dask/distributed v.2.21.0

**Comment by mrocklin** [comment]
Thank you for the comment @bachsh .  This issue is closed.   I recommend opening a new issue with more information so that maintainers can try to reproduce your issues.  https://blog.dask.org/2018/02/28/minimal-bug-reports
